Hot Reload Vue on shared/ changes

benjaminpjones opened the pull request:
Trying to figure out #28 (hot reload when shared/ changes)

While this was the one way I found to get hot reload to work, I don't think this solution is great for a couple of reasons.

- `"@/../../shared"` - it's pretty hacky to use relative imports.  It means we'll need to change a ton of files if we change the directory structure.
   - I'd prefer to import from `"@shared"` or something
   - I tried fiddling with "paths" and other similar config options, but no luck
- Would be better if we could depend on `dist` instead of `src`
    - I tried doing some of the magic using `watch.ignored` and `optimizeDeps` (see https://github.com/govariantsteam/govariants/tree/attempt-at-hot-reload) but wasn't able to turn on hot reload

That said, I am well beyond the limit of my TS and Vue/Vite knowledge, I am hoping someone might see a better path now that we have a "working" path

## Hot Reload Demo

Notice I am editing a variable in `shared/src/index.ts`!!!

https://github.com/govariantsteam/govariants/assets/25233703/8006b598-d154-4ac6-aee5-55581feb4762

JonKo314 commented:
> * it's pretty hacky to use relative imports
Yeah

> * Would be better if we could depend on `dist` instead of `src`
I'm not so sure about that one. `src` has all the typed stuff and doesn't need to be compiled, but it's just a feeling, I don't know enough about it.

It's very cool that you got it to work, but it would be sad if there wasn't a nicer way that just works.

benjaminpjones commented:
> > Would be better if we could depend on dist instead of src
> 
> I'm not so sure about that one. src has all the typed stuff and doesn't need to be compiled, but it's just a feeling, I don't know enough about it.

So I don't feel too strongly about this point compared to the first one.  However, my thought process is that `dist` is the directory that consumers would normally use when importing from the package (`"@ogfcommunity/variants-shared"`).

> it would be sad if there wasn't a nicer way that just works.

Indeed!  I'll need some help though - I've tried quite a few ways with very little luck 😞

benjaminpjones commented:
Okay actually, I think I got the absolute imports figured out.  Using `./shared` instead of `shared` in compilerOptions.paths helps

benjaminpjones commented:
Okay, I think I actually got this in a place that's publishable.

Summary of changes since earlier conversation:

- figured out how to keep the imports the same as before (not relative)
- I changed back to @ogfcommunity/variants-shared from @shared
    - not because I'm married to that name, but I wanted to keep focus on config changes in this PR
- Reverted tsconfig.vitest.json because it's of no consequence

benjaminpjones commented on packages/vue-client/src/views/HomeView.vue:
```diff
@@ -1,11 +1,12 @@
 <template>
   <main>
-    <h2>Welcome to Go variants!</h2>
+    <h2>Welcome to {{ SITE_NAME }}!</h2>
```
This was mainly for testing.  I can remove it before merging, but want to make sure we landed on an agreeable configuration first.

benjaminpjones commented on packages/vue-client/tsconfig.app.json:
```diff
@@ -4,9 +4,14 @@
+      "@ogfcommunity/variants-shared": ["./shared/src/index"]
     }
+    ,
+    "verbatimModuleSyntax": false,
```
I think `verbatimModuleSyntax` off makes it so I don't need to put `type { SomeImport }` **everywhere** in `shared/`.  I'm not opposed to it, but it would be quite a large change, and I wanted to keep this diff to config-only as much as possible.

benjaminpjones commented on packages/vue-client/tsconfig.app.json:
```diff
@@ -4,9 +4,14 @@
   "compilerOptions": {
     "composite": true,
-    "baseUrl": ".",
+    "baseUrl": "..",
```
This puts us in the `packages` directory, so we can look at `packages/shared`

benjaminpjones commented on packages/vue-client/tsconfig.config.json:
```diff
@@ -3,6 +3,14 @@
   "include": ["vite.config.*", "vitest.config.*", "cypress.config.*"],
```
I'm pretty sure the changes were required in all 3 tsconfig files (excluding tsconfig.vitest.json). Maybe there is a way to dedupe using the "extends" field.  But I don't quite understand the difference in intent behind `app` and `config` anyway...  Maybe we can remove them and just keep tsconfig.json?

benjaminpjones commented on packages/vue-client/vite.config.ts:
```diff
@@ -9,6 +9,12 @@ export default defineConfig({
     alias: {
       "@": fileURLToPath(new URL("./src", import.meta.url)),
+      "@ogfcommunity/variants-shared/src": fileURLToPath(
+        new URL("../shared/src", import.meta.url),
```
`@ogfcommunity/variants-shared/src` is to support imports like "@ogfcommunity/variants-shared/src/time_control"

`@ogfcommunity/variants-shared/src` is to support imports like "@ogfcommunity/variants-shared".

Importantly, the /src version needs to come first, else the latter will overtake it.
